src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearchWidget.ts

```diff
@@ -0,0 +1,276 @@
+		this.searchInput.setValue(options.value || '');
+		this.searchInput.setRegex(!!options.isRegex);
+		this.searchInput.setCaseSensitive(!!options.isCaseSensitive);
+		this.searchInput.setWholeWords(!!options.isWholeWords);
+		this._register(this.searchInput.inputBox.onDidChange(() => this.onSearchInputChanged()));
+		this._register(this.searchInput.inputBox.onDidHeightChange(() => this._onDidHeightChange.fire()));
+
+		this.searchInputFocusTracker = this._register(trackFocus(this.searchInput.inputBox.inputElement));
+		this._register(this.searchInputFocusTracker.onDidFocus(async () => {
+			this.searchInputBoxFocused.set(true);
+
+			const useGlobalFindBuffer = this.searchConfiguration.globalFindClipboard;
+			if (!this.ignoreGlobalFindBufferOnNextFocus && useGlobalFindBuffer) {
+				const globalBufferText = await this.clipboardServce.readFindText();
+				if (this.previousGlobalFindBufferValue !== globalBufferText) {
+					this.searchInput.inputBox.addToHistory();
+					this.searchInput.setValue(globalBufferText);
+					this.searchInput.select();
+				}
+
+				this.previousGlobalFindBufferValue = globalBufferText;
+			}
+
+			this.ignoreGlobalFindBufferOnNextFocus = false;
+		}));
+		this._register(this.searchInputFocusTracker.onDidBlur(() => this.searchInputBoxFocused.set(false)));
+	}
+
+	private onSearchInputChanged(): void {
+		this.searchInput.clearMessage();
+
+		if (this.searchConfiguration.searchOnType) {
+			if (this.searchInput.getRegex()) {
+				try {
+					const regex = new RegExp(this.searchInput.getValue(), 'ug');
+					const matchienessHeuristic = `
+								~!@#$%^&*()_+
+								\`1234567890-=
+								qwertyuiop[]\\
+								QWERTYUIOP{}|
+								asdfghjkl;'
+								ASDFGHJKL:"
+								zxcvbnm,./
+								ZXCVBNM<>? `.match(regex)?.length ?? 0;
+
+					const delayMultiplier =
+						matchienessHeuristic < 50 ? 1 :
+							matchienessHeuristic < 100 ? 5 : // expressions like `.` or `\w`
+								10; // only things matching empty string
+
+					this.submitSearch(true, this.searchConfiguration.searchOnTypeDebouncePeriod * delayMultiplier);
+				} catch {
+					// pass
+				}
+			} else {
+				this.submitSearch(true, this.searchConfiguration.searchOnTypeDebouncePeriod);
+			}
+		}
+	}
+
+	private submitSearch(triggeredOnType = false, delay: number = 0): void {
+		this.searchInput.validate();
+		if (!this.searchInput.inputBox.isInputValid()) {
+			return;
+		}
+
+		const value = this.searchInput.getValue();
+		const useGlobalFindBuffer = this.searchConfiguration.globalFindClipboard;
+		if (value && useGlobalFindBuffer) {
+			this.clipboardServce.writeFindText(value);
+		}
+		this._onSearchSubmit.fire({ triggeredOnType, delay });
+	}
+
+	private validateSearchInput(value: string): IMessage | null {
+		if (value.length === 0) {
+			return null;
+		}
+		if (!this.searchInput.getRegex()) {
+			return null;
+		}
+		try {
+			new RegExp(value, 'u');
+		} catch (e) {
+			return { content: e.message };
+		}
+
+		return null;
```

Comment:
Why null? Null really breaks a lot of things in code - I'd definitely recommend using undefined.

Reply:
@aaomidi As I mentioned in the PR description, notebookSearchView and notebookSearchWidget are almost of copies of searchView and searchWidget from vscode. Since I wasn't able to reuse them, I copied the necessary code and made modifications on top of them. Almost all of your comments refer to the code that I copied over. I'll go through them and see what I can do.